### ISSUE TITLE
ELLPACK for sparse multiply

### ISSUE THREAD

**Issue by atait** [nengo-labs/nengo-ocl#186]
_Advance notice of a pull request I am working on_. It is on branch atait/nengo-ocl:sparse-ellpack_impl

### Background
ELLPACK is an algorithm designed for GPUs. Currently, Nengo-OCL uses the CSR algorithm, which is better suited to CPUs. http://www.icl.utk.edu/files/publications/2014/icl-utk-772-2014.pdf

CSR weight indices are stored in a compressed format that is difficult to reconstitute in a non-serial way, which means that element multiply and accumulate-based reduction are performed by a single thread per row. Decompressing it relies on heavily non-coalesced memory access. 

ELL indices are stored in a padded array that can be accessed in parallel in a coalesced way. This means there can be one thread per elemental multiply, and a tree-based reduction occurs in local memory.

### Initial benchmark
ELLPACK implementation is more than 5x faster than Nengo-DL and Nengo-OCL-CSR in terms of SOPs/second.

![nengo-ocl-ellpack-100](https://user-images.githubusercontent.com/1455840/105022695-4b540300-5a07-11eb-85df-830e6c0548ee.png)

Some comments on the figure:
- Scipy (Reference), current Nengo-OCL (Nengo-OCL-csr) use CSR. I think Tensorflow (Nengo-DL) uses COO or CSR format
- The hockey stick on log-log axis means there is a constant-time launch-limited section, an elbow, and a polynomial-time synaptic operation (SOP)-limited section (SOP referring to both flops and memory access)
- Nengo-DL has some launch overhead. These simulators were warmed, and inference_only=True.
- Nengo-OCL-csr
    - has overhead similar to Reference (its warm time was longer though)
    - jumps super-linearly around 10^6. I don't know exactly why.
    - SOP performance approaches Nengo-DL, likely because they are being dominated by matvec and both using CSR
- Nengo-OCL-ell has the same launch overhead, has an elbow also around 10^6, but does not have that jump seen for OCL-CSR
- Nengo DL and OCL-ELL seems to support larger interconnects before memory allocation errors, but this is an artifact of the benchmark. For Watts-Strogatz, the fan-in is constant, meaning ELL format has its best case footprint. In general, CSR will have a more efficient footprint.

Benchmark details:
- Watts-Strogatz network with 100 fan-out per neuron
- GeForce RTX 2070 Super, 8GB
- nengo 3.1; nengo-DL 3.3.1.dev0
- nengo-ocl atait/nengo_ocl:sparse-ellpack_impl (based on 2.0.0.dev0)

#### Work group dimensioning
Currently all the groups are 1D with size related to the number of synapses per neuron, i.e. nonzero elements per row. I'm not sure this will always work with more than 1024 nonzero elements per row. The column matrix is reshaped to 1D before going in the RaggedArray, but it would be more natural to store it as 2D.

#### Occasional major bug
While it works for small (hundreds) and large (millions) interconnects, it sometimes fails for intermediate (thousands) sizes. The error is

```
...
nengo_ocl/plan.py, line 131, in call_n_times
    last_event.wait()
pyopencl._cl.LogicError: clWaitForEvents failed: <unknown error -9999>
```

This error does not occur when the number of neurons is a power of 2, so that's strange.

#### More tests
Current tests pass, but I think it would benefit from some more corner cases.

#### Clean up the batch script
called `benchmark_backends.py`. It makes it easier to compare different implementations, storing intermediate results in readable files. I think it will be useful in the future. It could go in a separate PR if that's easier.

#### Clean up commit history
Strip out unrelated changed I have made on that branch. Clean rebase upon nengo-labs master. Rewrite into meaningful commits. 

#### Make it default for `SparseDotInc`
After more testing

**Comment by atait** [comment]
Another benchmark using fan-in degree 1,000 instead of 100. There are 10x more synaptic operations compared to overhead and neuron operations, so there is no launch overhead visible, and ELLPACK does consistently better. The Reference is slightly super-linear. Perhaps the CPU cache hit rate is decreasing when less of the weight matrix fits in cache... just a guess.

![nengo-ocl-ellpack-1000](https://user-images.githubusercontent.com/1455840/105027233-be13ad00-5a0c-11eb-9c66-0421357fe050.png)

**Comment by atait** [comment]
I have hit all the TODO's above and made PR #187 (the benchmark script) and PR #188 (the ELLPACK implementation).

**Comment by hunse** [comment]
PR is merged, so I'm going to close this.
